genesis-tools/contracts/vesting-contract/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+    #[init]
+    pub fn new(
+        lockup_amount: u128,
+        lockup_timestamp: u64,
+        vesting_start_timestamp: u64,
+        vesting_cliff_timestamp: u64,
+        vesting_end_timestamp: u64,
+        owner_public_keys: Vec<PublicKey>,
+        foundation_public_keys: Vec<PublicKey>,
+    ) -> Self {
+        assert!(
+            near_bindgen::env::state_read::<VestingContract>().is_none(),
+            "The contract is already initialized"
+        );
+        let res = Self {
+            lockup_amount,
+            lockup_timestamp,
+            vesting_start_timestamp,
+            vesting_cliff_timestamp,
+            vesting_end_timestamp,
+            permanently_unstaked: false,
+        };
+        // It is okay for vesting start and vesting cliff to be in the past, but it does not
+        // makes sense to have lockup or vesting end to be in the past.
+        Self::check_timestamps_future(&[lockup_timestamp, vesting_end_timestamp]);
+        // The cliff should be between start and end of the vesting, potentially inclusive.
+        Self::check_timestamp_ordering(&[
+            vesting_start_timestamp,
+            vesting_cliff_timestamp,
+            vesting_end_timestamp,
+        ]);
+        // The lockup should be after start of the vesting, potentially inclusive.
+        Self::check_timestamp_ordering(&[vesting_start_timestamp, lockup_timestamp]);
+        let account_id = env::current_account_id();
+        for public_key in owner_public_keys {
+            Promise::new(account_id.clone()).add_access_key(
+                public_key,
+                0,
+                account_id.clone(),
+                KeyType::Owner.allowed_methods(),
+            );
+        }
+        for public_key in foundation_public_keys {
+            Promise::new(account_id.clone()).add_access_key(
+                public_key,
+                0,
+                account_id.clone(),
+                KeyType::Foundation.allowed_methods(),
+            );
+        }
+        res
+    }
+
+    /// Create a staking transaction on behalf of this account.
+    pub fn stake(&self, amount: u128, public_key: PublicKey) {
+        Self::assert_self();
+        assert!(!self.permanently_unstaked, "The account was permanently unstaked.");
+        assert!(
+            amount <= env::account_balance() + env::account_locked_balance(),
+            "Not enough balance to stake."
+        );
+        Promise::new(env::current_account_id()).stake(amount, public_key);
+    }
+
+    /// Get the amount of tokens that were vested.
+    pub fn get_unvested(&self) -> u128 {
+        let block_timestamp = env::block_timestamp();
+        if block_timestamp < self.vesting_cliff_timestamp {
+            self.lockup_amount
+        } else if block_timestamp >= self.vesting_end_timestamp {
+            0
+        } else {
+            let time_left = U256::from(self.vesting_end_timestamp - block_timestamp);
```

Comment:
```suggestion
            // cannot overflow since block_timestamp >= self.vesting_end_timestamp
            let time_left = U256::from(self.vesting_end_timestamp - block_timestamp); 
```